### PR TITLE
fix(storage): support mutation during insertion

### LINKED ISSUE #8108: bug: Failed to execute compact during insertion
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

0.8.67

### What's Wrong?

If there is an `insert/streaming load` operation during compression, the `optimize table xx compact` will fail.

```
MySQL [(none)]> optimize table xx compact;
ERROR 1105 (HY000): Code: 2009, displayText = TableVersionMismatched: 276953 expect `== 277527` but `277528`  while `update_table_meta`.
```

### How to Reproduce?

1. execute `optimize table xx compact`
2. execute insert/streaming load
3. compact failed and report error

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support compact、recluster and delete during insert.

In order to quickly fix bug, the code will be refactored sooner or later.

Fixes #8108

### CODE REVIEW

File: src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs [hunk 1]
```diff
@@ -222,24 +217,37 @@ impl TableMutator for ReclusterMutator {
...
-            table_info,
-            &self.base_mutator.location_generator,
-            new_snapshot,
-            &self.data_accessor,
-        )
-        .await?;
-        Ok(())
+        let table = FuseTable::try_from_table(table.as_ref())?;
+        match table
+            .commit_mutation(
```
Comment (Xuanwo) [inline src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs:236]: How about using `map_err` instead?

File: src/query/storages/fuse/src/operations/mutation/compact_mutator.rs [hunk 2]
```diff
@@ -173,20 +177,37 @@ impl TableMutator for CompactMutator {
...
-            new_snapshot,
-            &self.data_accessor,
-        )
-        .await
+        new_segments.append(&mut segments);
+        summary = merge_statistics(&summary, &merged_summary)?;
+
+        let table = FuseTable::try_from_table(table.as_ref())?;
+
+        match table
```
Comment (Xuanwo) [inline src/query/storages/fuse/src/operations/mutation/compact_mutator.rs:197]: The same.

File: src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs (hunk 1 above)
Comment (zhyass) [reply src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs:236]: > How about using `map_err` instead?

Contain async codes. Replace with `if let Err(e) = ... `.

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
Some code refactoring of fuse io files in this PR. Please update with the main branch:
https://github.com/datafuselabs/databend/pull/8262/files

File: src/query/storages/fuse/src/operations/commit.rs [hunk 3]
```diff
@@ -393,6 +395,59 @@ impl FuseTable {
...
+            }
+            new_segments.extend(segments.clone());
+            segments = new_segments;
+        }
+
+        let mut new_snapshot = TableSnapshot::from_previous(&latest_snapshot);
+        new_snapshot.segments = segments;
+        new_snapshot.summary = summary;
+
+        Self::commit_to_meta_server(
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/commit.rs:442]: shall we retry the commit if version-mismatched error is detected, and the refreshed snapshot is appended only?
there are cases the same table is kept being appended concurrently at a not-that-low frequency.

File: src/query/storages/fuse/src/operations/commit.rs [hunk 4]
```diff
@@ -399,54 +401,93 @@ impl FuseTable {
...
+                                "mutation conflicts, concurrent mutation detected while committing segment compaction operation",
+                            ));
+                        }
+
+                        // Check if there is only insertion during the operation.
+                        let mut new_segments = latest_snapshot.segments.clone();
+                        let suffix = new_segments.split_off(
+                            latest_snapshot.segments.len() - base_snapshot.segments.len(),
+                        );
+                        if suffix.ne(&base_snapshot.segments) {
```
Comment (dantengsky) [inline src/query/storages/fuse/src/operations/commit.rs:455]: do we need to call `abort_operation` here?

**Review by dantengsky — APPROVED** [review]
LGTM

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**